### DISCUSSION TITLE
How to get upload progress using fetch?

### DISCUSSION THREAD

**Discussion by wannacfuture** [wannacfuture/PsW#3] [Q&A] [answered]
I can't find the way to get the upload progress when I'm using fetch.

Seems, it doesn't support that feature at all.
But I can't use other(Axios, etc).

**Comment by sweetapplepumpkinpie** [comment, marked as answer]
Yes, you can track the progress of a file upload using the fetch API in JavaScript by using a custom ReadableStream to monitor the progress. Here's an example of how to do this:
async function uploadFile(file) {
  const url = 'YOUR_UPLOAD_URL';


  const onProgress = (loaded, total) => {
    const percentage = (loaded / total) * 100;
    console.log(`Progress: ${percentage.toFixed(2)}%`);
  };


  const monitorProgress = (reader, controller) => {
    const pushProgress = async () => {
      const { done, value } = await reader.read();
      if (done) {
        controller.close();
        return;
      }


      onProgress(value.byteLength, file.size);
      controller.enqueue(value);
      pushProgress();
    };


    pushProgress();
  };


  const progressStream = new ReadableStream({
    start(controller) {
      const reader = file.stream().getReader();
      monitorProgress(reader, controller);
    },
  });


  const response = await fetch(url, {
    method: 'PUT',
    body: progressStream,
  });


  if (!response.ok) {
    throw new Error('Upload failed');
  }
}


const fileInput = document.querySelector('input[type="file"]');
fileInput.addEventListener('change', (event) => {
  const file = event.target.files[0];
  uploadFile(file);
});
